Improve sentry frontend util

pgalic96 opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #2527 

#### Changes
<!-- What are the changes made in this pull request? -->

- Add version info
- Add sourcemaps in production
- Add `redux-sentry-middleware`

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@kschiffer Any other actions that need filtering in console?

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

kschiffer commented on config/webpack.config.babel.js:
```diff
@@ -150,6 +150,7 @@ export default {
     path: path.resolve(context, PUBLIC_DIR),
     crossOriginLoading: 'anonymous',
     publicPath: ASSETS_ROOT,
+    sourceMapFilename: production ? '[name].[chunkhash].js.map' : '[name].js.map',
```
So what are we doing now with the source map? It definitely shouldn't lay open for everyone in the public folder.

kschiffer commented on pkg/webui/console/store/index.js:
```diff
@@ -12,20 +12,88 @@
 import requestPromiseMiddleware from './middleware/request-promise-middleware'
 import logics from './middleware/logics'
 
+if (env.sentryDsn)
```
That shouldn't be necessary, no? Sentry was already initialized in `console.js`. There should be away to get the already instantiated Sentry object.

kschiffer replied:
Also let's not use non-bracketed if statements when followed by a mult-line statement. Generally, I think we should use non-bracketed if statements only sparsely. I'm ok with using then if they are e.g. followed by a simple return statement.

kschiffer commented on pkg/webui/console/store/index.js:
```diff
@@ -12,20 +12,88 @@
 export default function(history) {
   const middleware = applyMiddleware(
+    createSentryMiddleware(Sentry, {
+      actionTransformer: action => {
```
Please move this to a separate module. I also sense that there might be a library that simplifies scrapping keywords from an object. Can you check that?

kschiffer commented on pkg/webui/oauth/store/index.js:
```diff
@@ -12,19 +12,58 @@
-  const middleware = applyMiddleware(routerMiddleware(history), createLogicMiddleware(logic))
+  const middleware = applyMiddleware(
+    createSentryMiddleware(Sentry, {
+      actionTransformer: action => {
```
Idem.

kschiffer commented on pkg/webui/console/store/index.js:
```diff
@@ -12,20 +12,88 @@
+            ...state.user,
+            user: {
+              ...state.user.user,
+              ids: undefined,
```
Sentry offers a way to remove the user id while retaining a correlatable ID, see: https://docs.sentry.io/data-management/sensitive-data/#examples

pgalic96 replied:
in relation to this, `user_id` does not seem to be Personally Identifiable Information, so I don't think we should scrub this.

kschiffer replied:
Not per se but it can be,
1. if e.g. a user chooses to use his/her full name as ID
2. once it is correlated to our user account table

pgalic96 commented:
@kschiffer I think it's ready for review. In order to test this the following setup is necessary:
- Create an auth token for our organization
- Fill in required fields in `sentry.properties` file
- Set an environment variable `SENTRY_PROPERTIES=$PWD/config/sentry.properties`
- Add a random error somewhere
- Build a production build (I suggest to run `rm -rf node_modules sdk/js/node_modules .cache yarn.lock sdk/js/yarn.lock` and then `mage jsSDK:clean jsSDK:deps jsSDK:build js:clean js:deps js:build`)
- Observe stack trace in our sentry 

Possible regressions:
- `public` folder is not cleaned after the build
- A production build should be built successfully by just using `mage js:build` and maybe `mage js:clean` before. If it is necessary to run the `rm -rf node_modules...` every time then that is definitely a regression since the previous build. This can be checked by checking if the files generated by `mage js:build` are the same as the ones in the manifest. Also, sentry should not complain about missing sourcefiles.

kschiffer commented on config/sentry.properties:
```diff
@@ -0,0 +1,5 @@
+defaults.org=
```
In CI we'll have to source the values from the environment.

https://docs.sentry.io/cli/configuration/#properties-files leaves me a bit unclear wether all of these can simply be set as env variables as well. Can you check that? In any case, it does not make sense to check-in this file.

kschiffer commented on config/webpack.config.babel.js:
```diff
@@ -98,7 +100,7 @@ export default {
   stats: 'minimal',
   target: 'web',
-  devtool: production ? false : 'eval-source-map',
+  devtool: production ? 'source-map' : 'eval-source-map',
```
I see now that we produce source files for all JS and CSS files in the output, including all of the locale files. I'm sure there's a way to limit this to a whitelist only (should be just the main bundle's JS and CSS).

pgalic96 replied:
I don't think we should produce sourcemaps for css files as well, it's not useful for sentry

kschiffer commented on pkg/webui/constants/sensitive-data.js:
```diff
@@ -0,0 +1,15 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export default ['primary_email_address', 'eui', 'apiKeys', 'dev_eui', 'join_eui']
```
@htdvisser can you share a braindump of what other properties you deem to be sensitive?

```suggestion
export default ['primary_email_address', 'email', 'email_address', 'contact_info', 'eui', 'apiKeys', 'api_keys', 'dev_eui', 'join_eui', 'password', 'keys', 'f_nwk_s_int_key', 's_nwk_s_int_key', 'nwk_s_enc_key', 'app_s_key', 'attributes']
```

It would be great if it was possible to define regexps also, e.g. `/key/`.

htdvisser replied:
Sentry is compliant with GDPR and other data protection laws, so the goal is not necessarily to redact all privacy-sensitive information. The goal is to redact *secret* information.

I think it's good to make this a regex indeed, then we can filter out:

- equal `password` or end with `.password` or `_password`
- equal `key` or end with `.key` or `_key`
- equal `secret` or end with `.secret` or `_secret`

I think it's a bad idea to redact IDs or EUIs, because that would prevent us from adequately debugging.

kschiffer commented on pkg/webui/constants/sentry.js:
```diff
@@ -0,0 +1,29 @@
+
+import env from '@ttn-lw/lib/env'
+
+const sentryConfig = {
```
Can you also look at adding additional context data: https://docs.sentry.io/enriching-error-data/additional-data/?platform=browser

I think, especially `Sentry.setUser()` would be useful. Based on the action or current location of the user we could set tags like `application`, `gateway,` `end-device`, and even action types such as `get`, `list`, `update`, `delete`, `create` using `Sentry.setTag()`. For TTES we could set tenant and cluster as well.

pgalic96 replied:
Yes, indeed this would be very nice. I have set the `user_id` tag in the store. I really like the rest of your ideas, however it is a bit more tricky since the `getTag` function of `redux-sentry-middleware` supports passing only `state`, therefore action type or payload will have to be injected some other way in the tags.

I already have ideas for adding action type tags, and tags based on location, but given the priority of this issue I think this should be added in another PR. I suggest we go ahead with this (and it will probably take some time to test that everything even without the tags).

pgalic96 commented:
@kschiffer Fixed most of the feedback, see individual comment for the rest.

You were right, it is possible to configure this with env variables, in order to have `sentry-cli` working properly the following env variables need to be set:

```
export SENTRY_ORG=
export SENTRY_PROJECT=
export SENTRY_AUTH_TOKEN=
export SENTRY_DSN=
```

kschiffer commented on config/webpack.config.babel.js:
```diff
@@ -47,6 +48,7 @@ const WEBPACK_DEV_SERVER_USE_TLS = process.env.WEBPACK_DEV_SERVER_USE_TLS === 't
 const TTN_LW_TLS_CERTIFICATE = process.env.TTN_LW_TLS_CERTIFICATE || './cert.pem'
 const TTN_LW_TLS_KEY = process.env.TTN_LW_TLS_KEY || './key.pem'
 const TTN_LW_TLS_ROOT_CA = process.env.TTN_LW_TLS_ROOT_CA || './cert.pem'
+const USE_SENTRY_SOURCEMAPS = Boolean(process.env.SENTRY_DSN)
```
```suggestion
const WEBPACK_UPLOAD_SENTRY_SOURCEMAPS = Boolean(process.env.SENTRY_DSN) && process.env.WEBPACK_UPLOAD_SENTRY_SOURCEMAPS === 'true'
const TTN_LW_TLS_CERTIFICATE = process.env.TTN_LW_TLS_CERTIFICATE || './cert.pem'
const TTN_LW_TLS_KEY = process.env.TTN_LW_TLS_KEY || './key.pem'
const TTN_LW_TLS_ROOT_CA = process.env.TTN_LW_TLS_ROOT_CA || './cert.pem'
```

I think it's better to set this explicitly, otherwise we might end up accidentally uploading faulty source maps from local machines.

htdvisser replied:
Just noticed this stuff for the first time. I don't remember any documentation saying that webpack config only supports environment variables (and not the config file)

kschiffer replied:
Hmm. That's true. These are all development related though (changing behavior of the dev server or CI) and none of those settings are needed for production. 

Would it be acceptable for you if we document this constraint, or should we take time to source the config file as well?

htdvisser replied:
I'm fine with documenting it

kschiffer replied:
@pgalic96 can you take care as part of updating the development documentation for the new environment variables? A simple sentence stating that these configs cannot be source from the config file should suffice.

pgalic96 commented:
Added webpack config info in the `DEVELOPMENT.md` and also a `Sentry` section in the documentation. I added only the part about uploading sourcemaps to keep it scoped to this PR. cc @htdvisser @kschiffer

kschiffer commented on pkg/webui/lib/omit_test.js:
```diff
@@ -0,0 +1,201 @@
+      expect(result.apiKeys).toBeUndefined()
+      expect(result.pagination.applications.ids).toBeUndefined()
+      expect(result.pagination.gateways.ids).toBeUndefined()
+      expect(result.pagination.apiKeys).toBeUndefined()
```
This still holds true. The whole result object could consist of undefined values and the test would still pass.

pgalic96 replied:
Lines 192 and 193 should take care of this now

kschiffer commented on DEVELOPMENT.md:
````diff
@@ -211,6 +211,8 @@ TTN_LW_IS_EMAIL_NETWORK_IDENTITY_SERVER_URL="http://localhost:8080/oauth.js"
 TTN_LW_CONSOLE_UI_ASSETS_BASE_URL="http://localhost:8080/assets"
 ```
 
+> Note: Development configuration can be loaded from environment variables only. It cannot be sourced from a config file.
+
````
This is in the wrong place. The configuration above can very well be source from the config file. This line needs to be at L217 and say something like 
> Note: webpack-related configuration can be loaded from environment variables only. It cannot be sourced from a config file.

kschiffer commented on CHANGELOG.md:
```diff
@@ -9,6 +9,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Redux state and actions reporting to Sentry
+- Serving frontend sourcemaps in production
```
Ending periods.
